Cython/Utility/Exceptions.c

```diff
@@ -743,29 +743,30 @@ static PyCodeObject* __Pyx_CreateCodeObjectForTraceback(
             const char *funcname, int c_line,
             int py_line, const char *filename) {
     PyCodeObject *py_code = 0;
-    PyObject *py_srcfile = 0;
     PyObject *py_funcname = 0;
-
     #if PY_MAJOR_VERSION < 3
+    PyObject *py_srcfile = 0;
+
     py_srcfile = PyString_FromString(filename);
-    #else
-    py_srcfile = PyUnicode_FromString(filename);
-    #endif
     if (!py_srcfile) goto bad;
+    #endif
+
     if (c_line) {
         #if PY_MAJOR_VERSION < 3
         py_funcname = PyString_FromFormat( "%s (%s:%d)", funcname, $cfilenm_cname, c_line);
         #else
         py_funcname = PyUnicode_FromFormat( "%s (%s:%d)", funcname, $cfilenm_cname, c_line);
+        if (!py_funcname) goto bad;
+        funcname = PyUnicode_AsUTF8(py_funcname);
+        if (!funcname) goto bad;
         #endif
     }
     else {
         #if PY_MAJOR_VERSION < 3
         py_funcname = PyString_FromString(funcname);
-        #else
-        py_funcname = PyUnicode_FromString(funcname);
         #endif
     }
+    #if PY_MAJOR_VERSION < 3
     if (!py_funcname) goto bad;
```

Comment:
This can be moved into the `else` case now.

Reply:
It was covering two cases (which is why I left it there). With that said, it's probably clearer attached to each case individually, so I moved it.